honeybadgermpc/hbavss_light.py

```diff
@@ -0,0 +1,135 @@
+        def multicast(msg):
+            for i in range(self.n):
+                self.send(i, msg)
+
+        commitments, ephemeral_public_key, encrypted_witnesses = loads(avss_msg)
+        shared_key = pow(ephemeral_public_key, self.private_key)
+        share, witness = SymmetricCrypto.decrypt(
+            str(shared_key).encode("utf-8"), encrypted_witnesses[self.my_id])
+        if self.poly_commit.verify_eval(commitments, self.my_id+1, share, witness):
+            multicast((avss_id, HbAVSSMessageType.OK))
+        else:
+            logging.error("PolyCommit verification failed.")
+            raise HoneyBadgerMPCError("PolyCommit verification failed.")
+
+        oks_recvd = 0
+        while oks_recvd < 2*self.t + 1:
+            _, avss_msg = await recv()  # First value is the `sid`
+            if avss_msg[1] == HbAVSSMessageType.OK:
+                oks_recvd += 1
+
+        logging.debug("[%d][HbAVSSLight] 2t+1 OKs received.", self.my_id)
+        return share
+
+    def _get_dealer_msg(self, value):
+        phi = self.poly.random(self.t, value)
+        commitments, aux_poly = self.poly_commit.commit(phi)
+        ephemeral_secret_key = self.field.random()
+        ephemeral_public_key = pow(self.g, ephemeral_secret_key)
+        z = [None]*self.n
+        for i in range(self.n):
+            witness = self.poly_commit.create_witness(aux_poly, i+1)
+            shared_key = pow(self.public_keys[i], ephemeral_secret_key)
+            z[i] = SymmetricCrypto.encrypt(
+                str(shared_key).encode("utf-8"), (phi(i+1), witness))
+
+        return dumps((commitments, ephemeral_public_key, z))
+
+    async def avss(self, value=None, dealer_id=None, client_mode=False):
+        # If `value` is passed then the node is a 'Sender'
+        # `dealer_id` must be equal to `self.my_id`
+        if value is not None:
+            if dealer_id is None:
+                dealer_id = self.my_id
+            assert dealer_id == self.my_id, "Only dealer can share a value."
+        # If `value` is not passed then the node is a 'Recipient'
+        # Verify that the `dealer_id` is not the same as `self.my_id`
+        elif dealer_id is not None:
+            assert dealer_id != self.my_id
+        if client_mode:
+            assert dealer_id is not None
+            assert dealer_id == self.n
+
+        # Choose the avss_id based on the order this is called
+        avss_id = f"AVSS-{self.counter}"
+        self.counter += 1
+        logging.debug("[%d] Starting AVSS [%s]", self.my_id, avss_id)
+
+        broadcast_msg = None if self.my_id != dealer_id else self._get_dealer_msg(value)
+        n = self.n if not client_mode else self.n+1
+        logging.debug("N: %d %d", n, self.n)
+
+        avss_msg = await reliablebroadcast(
+            avss_id,
+            self.my_id,
+            n,
+            self.t,
+            dealer_id,
+            broadcast_msg,
+            self.queues[avss_id].get,
```

Comment:
There is a potential message ordering problem here. You are using the same read channel `self.queues[avss_id].get` for both the reliable broadcast subprotocol and for the AVSS-specific OK messages. It is possible that these two kinds of messages can be interleaved. These will need to be partitioned somehow into, for example, `self.RBCqueues` and `self.AVSSqueues` based on a tag

Reply:
Makes sense. If someone finishes RBC and starts sending AVSS OK messages then that will screw things up. Will split this into avss and rbc queues.

Reply:
This has been fixed.